docs/design.md

```diff
@@ -0,0 +1,156 @@
+like zypper.
+
+## How Hypper Differs From Helm
+
+Why build a package manager on a package manager? Helm is a package manager that
+tries to make few assumptions. For example, Custom Resource Definitions (CRD)s
+are a Kubernetes resource that extend the Kubernetes API. They can be used in
+various ways and there are pitfalls with them. Helm makes very few assumptions
+when it comes to them. Hypper is going to make some different assumptions which
+will result in different features.
+
+As differences between Hypper and Helm arise they will be documented along with
+the reasons for the difference. This is in an effort to provide clarity.
+
+When features from Hypper can be upstreamed into Helm they will be. In that way,
+Hypper can act as a testing ground for Helm. We do not expect that every feature
+will fit in upstream Helm.
+
+A second element in the difference is that Helm, as a package manager, does not
+handle environment deployment operations. Those are left to higher level tools
+like Helmfile (a push model like Ansible) and the Flux Helm Operator (a pull
+model like Chef). Helm expects higher level tools will use Helm and charts for
+operating in different environments.
+
+Hypper will move will make some assumptions that move it more towards the
+environment management. This is in a different way from Helmfile or the Flux
+Helm Operator but further than Helm will go.
+
+## Directory Structure
+
+The directory structure is inspired by Helm with the following layout.
+
+```
+.
+├── cmd
+│   └── hypper
+├── docs
+└── pkg
+    └── action 
+    └── ... 
+```
+
+- The _cmd/hypper_ directory contains the client application. This CLI for mac,
+  linux, and windows is both a functional application and reference application
+  of the library.
+- _docs_ contains documentation on the design and usage of hypper.
+- The _pkg_ directory contains a library meant to be imported by _cmd/hypper_ and
+  other applications. The functionality needed is implemented here.
+- _pkg/actions_ contains the main functionality used by _cmd/hypper_.
+
+This is the same layout as Helm itself.
+
+## Design Intent
+
+The following outline the initial major intentions of hypper.
+
+### Modern Terminal UI
+
+Colors and icons are often present in modern terminals. Sometimes referred to
+as eye candy, these features can improve user experience by drawing attention to
+important things.
+
+### Optional Dependencies
+
+In Helm there are two ways to handle dependencies.
+
+1. Explicitly declare other charts as dependencies. This will cause those charts
+   to be installed. When the templates are rendered they happen in the same
+   sandboxed environment.
+2. Install charts separately and use values to connect them to each other. For
+   example, install PostreSQL with one chart and then pass connection information
+   into another chart using a Secret or values that generate a Secret.
+
+Hypper intends to add a 3rd form of dependency via an optional dependency. For
```

Comment:
Do we expect this to be used?

Like I can understand the reasoning behind it, but that means that you are not testing the same deployment on dev vs prod and that your values.yaml have to be different as well (to take care of disabling that chart deployment)

Is hypper expected to automatically disable that chart if the dependencies are optional and not installed or does that fall into the end user to do? How do we plan to account for the errors/issues that can come from that mismatch?

Reply:
> Is hypper expected to automatically disable that chart if the dependencies are optional and not installed or does that fall into the end user to do?

I side with a user option (reminds me of `Recommends` and `Suggests` package relations in .deb, that can be optionally installed if configured to do so).

> How do we plan to account for the errors/issues that can come from that mismatch?

We could solve for all the charts to be installed, before installing anything on the cluster, for their APIs inputs and outputs. And error if there can't be met.

Reply:
> > Is hypper expected to automatically disable that chart if the dependencies are optional and not installed or does that fall into the end user to do?
>
> I side with a user option (reminds me of Recommends and Suggests package relations in .deb, that can be optionally installed if configured to do so).

This is inspired by other package managers.

hypper and Helm are package managers rather than something like Helmfile, Ansible, Chef, or the Flux Helm Operator. it's a different concern.

Sometimes a local development environment can't run all the things. We are loading our clusters up with lots of services. On my local 16gb laptop I can't run everything. Sometimes I'll want a minimal install for dev. I would not consider my laptop running part of something to be the test that it's ready to go to prod. Someone should have another test.

Just like Helm, it's a tool and it's on users to make their own decisions.

I expect the install to be interactive and to disable that with flags that choose options. We will want smart defaults.

> How do we plan to account for the errors/issues that can come from that mismatch?

We prompt end users. Or, if in non-interactive mode we error out as early as we can. We want to be as safe as possible.